File: ProtocJsGenPlugin/TypeHelper.cs
using Google.Protobuf.Reflection;
using System.Linq;

namespace ProtocJsGenPlugin;

internal static class TypeHelper
{
    public static string GetJsType(FieldDescriptorProto field, FileDescriptorProto protoFile)
    {
        var baseType = GetBaseJsType(field, protoFile);

        // 处理repeated字段（数组）
        if (field.Label == FieldDescriptorProto.Types.Label.Repeated)
        {
            // 检查是否是map（proto中的map会被表示为repeated的特定消息类型）
            if (IsMapField(field))
            {
                var mapTypes = GetMapKeyValueTypes(field);
                return $"Map<{mapTypes.keyType}, {mapTypes.valueType}>";
            }
            else
            {
                return $"{baseType}[]";
            }
        }

        // 处理optional字段（proto3中所有字段都是可选的，但我们可以标记为可能为null）
        // 对于标量类型，在JS中可以是undefined
        return baseType;
    }

    public static string GetBaseJsType(FieldDescriptorProto field, FileDescriptorProto protoFile)
    {
        switch (field.Type)
        {
            case FieldDescriptorProto.Types.Type.Double:
            case FieldDescriptorProto.Types.Type.Float:
                return "number";

            case FieldDescriptorProto.Types.Type.Int64:
            case FieldDescriptorProto.Types.Type.Uint64:
            case FieldDescriptorProto.Types.Type.Int32:
            case FieldDescriptorProto.Types.Type.Fixed64:
            case FieldDescriptorProto.Types.Type.Fixed32:
            case FieldDescriptorProto.Types.Type.Uint32:
            case FieldDescriptorProto.Types.Type.Sfixed32:
            case FieldDescriptorProto.Types.Type.Sfixed64:
            case FieldDescriptorProto.Types.Type.Sint32:
            case FieldDescriptorProto.Types.Type.Sint64:
                return "number";

            case FieldDescriptorProto.Types.Type.Bool:
                return "boolean";

            case FieldDescriptorProto.Types.Type.String:
                return "string";

            case FieldDescriptorProto.Types.Type.Bytes:
                return "Uint8Array";

            case FieldDescriptorProto.Types.Type.Enum:
                return field.TypeName.Split('.').Last();

            case FieldDescriptorProto.Types.Type.Message:
                return GetMessageTypeName(field.TypeName, protoFile);

            default:
                return "any";
        }
    }

    public static bool IsMapField(FieldDescriptorProto field)
    {
        // 简单判断：如果是repeated消息类型，并且类型名包含"Entry"，可能是map
        // 更准确的实现需要检查消息是否有key和value字段
        return field.Type == FieldDescriptorProto.Types.Type.Message &&
               field.Label == FieldDescriptorProto.Types.Label.Repeated &&
               field.TypeName?.Contains("Entry") == true;
    }

    public static (string keyType, string valueType) GetMapKeyValueTypes(FieldDescriptorProto field)
    {
        // 简化实现：返回通用的类型
        // 实际应该解析消息结构获取key和value类型
        return ("string", "any");
    }

    public static string GetMessageTypeName(string typeName, FileDescriptorProto protoFile)
    {
        // 移除开头的"."，因为proto类型名以"."开头表示绝对路径
        if (typeName.StartsWith("."))
            typeName = typeName.Substring(1);

        // 检查是否是当前文件中的嵌套消息
        // 如果类型名包含点（除了包名之外的点），则可能是嵌套消息
        if (!string.IsNullOrEmpty(protoFile.Package) && typeName.StartsWith(protoFile.Package + "."))
        {
            // 去掉包名部分
            string withoutPackage = typeName.Substring(protoFile.Package.Length + 1);
            // 如果去掉包名后仍包含点，说明是嵌套消息
            if (withoutPackage.Contains('.'))
            {
                // 返回独立类名
                return GetIndependentClassName(typeName, protoFile);
            }
        }

        // 返回类型名（简化处理，实际可能需要处理嵌套）
        return typeName.Split('.').Last();
    }

    public static string GetIndependentClassName(string fullTypeName, FileDescriptorProto? protoFile = null)
    {
        // 移除前导点
        if (fullTypeName.StartsWith("."))
            fullTypeName = fullTypeName.Substring(1);

        // 移除包名前缀（如果提供了protoFile）
        if (protoFile != null && !string.IsNullOrEmpty(protoFile.Package) && fullTypeName.StartsWith(protoFile.Package + "."))
        {
            fullTypeName = fullTypeName.Substring(protoFile.Package.Length + 1);
        }

        // 将点替换为下划线
        string className = fullTypeName.Replace('.', '_');

        // 添加前缀以确保不与顶级类名冲突
        return "__" + className;
    }

    public static string GetMethodName(FieldDescriptorProto field)
    {
        return field.Name.SnakeToPascalCase();
    }
}

File: src/type_helper.cc
#include "type_helper.h"

#include <algorithm>
#include <cctype>
#include <string>

#include "string_extensions.h"

namespace protoc_js_gen_plugin {

namespace {

using google::protobuf::FieldDescriptorProto;
using google::protobuf::FileDescriptorProto;

// Helper to get the last component of a dot-separated string
std::string GetLastComponent(const std::string& str) {
    size_t pos = str.find_last_of('.');
    if (pos == std::string::npos) return str;
    return str.substr(pos + 1);
}

}  // namespace

// Static member initialization
TypeHelper::TypeNameTransformer TypeHelper::type_name_transformer_ = nullptr;

void TypeHelper::SetTypeNameTransformer(TypeNameTransformer transformer) {
    type_name_transformer_ = std::move(transformer);
}

TypeHelper::TypeNameTransformer TypeHelper::GetTypeNameTransformer() {
    return type_name_transformer_;
}

std::string TypeHelper::GetJsType(
    const FieldDescriptorProto& field,
    const FileDescriptorProto& proto_file) {

    std::string base_type = GetBaseJsType(field, proto_file);

    // Handle repeated fields (arrays)
    if (field.label() == FieldDescriptorProto::LABEL_REPEATED) {
        // Check if it's a map field
        if (IsMapField(field)) {
            // JSON serialization doesn't support Map type
            return "any";
        } else {
            return base_type + "[]";
        }
    }

    return base_type;
}

std::string TypeHelper::GetBaseJsType(
    const FieldDescriptorProto& field,
    const FileDescriptorProto& proto_file) {

    switch (field.type()) {
        case FieldDescriptorProto::TYPE_DOUBLE:
        case FieldDescriptorProto::TYPE_FLOAT:
            return "number";

        case FieldDescriptorProto::TYPE_INT64:
        case FieldDescriptorProto::TYPE_UINT64:
        case FieldDescriptorProto::TYPE_INT32:
        case FieldDescriptorProto::TYPE_FIXED64:
        case FieldDescriptorProto::TYPE_FIXED32:
        case FieldDescriptorProto::TYPE_UINT32:
        case FieldDescriptorProto::TYPE_SFIXED32:
        case FieldDescriptorProto::TYPE_SFIXED64:
        case FieldDescriptorProto::TYPE_SINT32:
        case FieldDescriptorProto::TYPE_SINT64:
            return "number";

        case FieldDescriptorProto::TYPE_BOOL:
            return "boolean";

        case FieldDescriptorProto::TYPE_STRING:
            return "string";

        case FieldDescriptorProto::TYPE_BYTES:
            return "Uint8Array";

        case FieldDescriptorProto::TYPE_ENUM:
            if (type_name_transformer_) {
                std::string transformed = type_name_transformer_(field.type_name(), proto_file);
                if (!transformed.empty()) {
                    return transformed;
                }
            }
            return GetLastComponent(field.type_name());

        case FieldDescriptorProto::TYPE_MESSAGE:
            if (type_name_transformer_) {
                std::string transformed = type_name_transformer_(field.type_name(), proto_file);
                if (!transformed.empty()) {
                    return transformed;
                }
            }
            return GetMessageTypeName(field.type_name(), proto_file);

        default:
            return "any";
    }
}

bool TypeHelper::IsMapField(const FieldDescriptorProto& field) {
    // Simple check: if it's a repeated message type and type name contains "Entry"
    return field.type() == FieldDescriptorProto::TYPE_MESSAGE &&
           field.label() == FieldDescriptorProto::LABEL_REPEATED &&
           field.type_name().find("Entry") != std::string::npos;
}

std::pair<std::string, std::string> TypeHelper::GetMapKeyValueTypes(
    const FieldDescriptorProto& field) {
    // Simplified implementation
    return {"string", "any"};
}

std::string TypeHelper::GetMessageTypeName(
    const std::string& type_name,
    const FileDescriptorProto& proto_file) {

    // Remove leading "." if present
    std::string processed_name = type_name;
    if (!processed_name.empty() && processed_name[0] == '.') {
        processed_name = processed_name.substr(1);
    }

    // Check if it's a nested message in the current file
    if (!proto_file.package().empty() &&
        processed_name.find(proto_file.package() + ".") == 0) {

        // Remove package prefix
        std::string without_package = processed_name.substr(proto_file.package().length() + 1);

        // If still contains dots, it's a nested message
        if (without_package.find('.') != std::string::npos) {
            return GetIndependentClassName(type_name, &proto_file);
        }
    }

    return GetLastComponent(processed_name);
}

std::string TypeHelper::GetIndependentClassName(
    const std::string& full_type_name,
    const FileDescriptorProto* proto_file) {

    // Remove leading dot
    std::string processed_name = full_type_name;
    if (!processed_name.empty() && processed_name[0] == '.') {
        processed_name = processed_name.substr(1);
    }

    // Remove package prefix if proto_file is provided
    if (proto_file && !proto_file->package().empty()) {
        std::string package_prefix = proto_file->package() + ".";
        if (processed_name.find(package_prefix) == 0) {
            processed_name = processed_name.substr(package_prefix.length());
        }
    }

    // Replace dots with underscores
    std::string class_name;
    for (char c : processed_name) {
        if (c == '.') {
            class_name += '_';
        } else {
            class_name += c;
        }
    }

    // Add prefix to avoid conflicts with top-level class names
    return "__" + class_name;
}

std::string TypeHelper::GetMethodName(
    const FieldDescriptorProto& field) {

    return SnakeToPascalCase(field.name());
}

}  // namespace protoc_js_gen_plugin